File: ChartTable/ChartTable/ViewController.m
//
//  ViewController.m
//  ChartTable
//
//  Created by xun.liu on 15/5/25.
//  Copyright (c) 2015年 xun.liu. All rights reserved.
//

#import "ViewController.h"
#import "DataTableView.h"
#import "JSONKit.h"

@interface ViewController ()<scrollDelegate, UIScrollViewDelegate>

@end

@implementation ViewController
{
    NSArray *_arrX;
    NSMutableArray *_arrY;
    UIScrollView *_scroll;
    UIView *_headView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    NSString *path = [[NSBundle mainBundle] pathForResource:@"data" ofType:@"json"];
    NSString *str = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
    NSDictionary *dic = [str objectFromJSONString];
    NSDictionary *dataDic = [dic objectForKey:@"data"];
    
    //x方向数据
    _arrX = [dataDic objectForKey:@"table_titles"];
    
    //y方向数据
    _arrY = [dataDic objectForKey:@"list_datas"];
    
    
    //第一列tableView父视图
    _headView = [[UIView alloc] initWithFrame:CGRectMake(30, 50, 200, 300)];
    _headView.backgroundColor = [UIColor grayColor];
    _headView.userInteractionEnabled = YES;
    //设置边框，形成表格
    _headView.layer.borderWidth = .5f;
    _headView.layer.borderColor = [UIColor blackColor].CGColor;
    [self.view addSubview:_headView];
    
    //可左右滑动tableView父视图
    _scroll = [[UIScrollView alloc] initWithFrame:CGRectMake(230, 50, 500, 300)];
    _scroll.contentSize = CGSizeMake((_arrX.count-1)*130, _arrY.count*50);
    _scroll.showsHorizontalScrollIndicator = NO;
    _scroll.showsVerticalScrollIndicator = NO;
    _scroll.bounces = NO;
    _scroll.delegate = self;
    _scroll.backgroundColor = [UIColor lightGrayColor];
    //设置边框，形成表格
    _scroll.layer.borderWidth = .5f;
    _scroll.layer.borderColor = [UIColor blackColor].CGColor;
    
    [self.view addSubview:_scroll];
    
    
    //第一列表
    DataTableView *tableView = [[DataTableView alloc] initWithFrame:_headView.bounds style:UITableViewStylePlain];
    
    NSMutableArray *titleArr1 = [NSMutableArray array];
    NSMutableArray *levels = [NSMutableArray array];
    for (NSDictionary *dic in _arrY)
    {
        NSString *titleStr = [dic objectForKey:@"date1"];
        [titleArr1 addObject:titleStr];
        
        NSString *level = [dic objectForKey:@"table_level"];
        [levels addObject:level];
    }
    tableView.titleArr = titleArr1;
    tableView.levels = levels;
    
    tableView.headerStr = [_arrX[0] objectForKey:@"name1"];
    tableView.scroll_delegate = self;
    [_headView addSubview:tableView];
    
    
    for (int i=0; i<_arrX.count-1; i++)
    {
        DataTableView *tableView = [[DataTableView alloc] initWithFrame:CGRectMake(130*i, 0, 130, 300) style:UITableViewStylePlain];
        
        //x方向 取出key对应的字符串名字
        NSString *xkey = [NSString stringWithFormat:@"name%i",i+2];
        NSString *xname = [_arrX[i+1] objectForKey:xkey];
        tableView.headerStr = xname;
        
        //y方向
        NSMutableArray *titleArr2 = [NSMutableArray array];
        for (int j=0; j<_arrY.count; j++)
        {
            NSString *ykey = [NSString stringWithFormat:@"date%i",i+2];
            NSString *yname = [_arrY[j] objectForKey:ykey];
            [titleArr2 addObject:yname];
        }
        tableView.titleArr = titleArr2;
        [tableView reloadData];
        tableView.scroll_delegate = self;
        [_scroll addSubview:tableView];
    }
}

#pragma mark - scrollDelegate
-(void)dataTableViewContentOffSet:(CGPoint)contentOffSet
{
    for (UIView *subView in _scroll.subviews)
    {
        if ([subView isKindOfClass:[DataTableView class]])
        {
            [(DataTableView *)subView setTableViewContentOffSet:contentOffSet];
        }
    }
    
    for (UIView *subView in _headView.subviews)
    {
        [(DataTableView *)subView setTableViewContentOffSet:contentOffSet];
    }
}

#pragma mark - UIScrollViewDelegate
-(void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGPoint p = scrollView.contentOffset;
    NSLog(@"%@",NSStringFromCGPoint(p));
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: ChartTable_Swift/ChartTable_Swift/ViewController.swift
//
//  ViewController.swift
//  ChartTable_Swift
//
//  Created by xun.liu on 16/5/23.
//  Copyright © 2016年 TVM. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    var arrX:Array = [JSON]()
    var arrY:Array = [JSON]()
    var scrollView: UIScrollView?
    var headerView: UIView?
    var dataJson:JSON = JSON.null
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let path = Bundle.main.path(forResource: "data", ofType: "json")
        let data = try? Data(contentsOf: URL(fileURLWithPath: path!))
        guard let json = try? JSON(data: data!) else {
            return
        }
        dataJson = json
        
        print("dataJson = \(dataJson)")
        arrX = dataJson["data"]["table_titles"].arrayValue
        arrY = dataJson["data"]["list_datas"].arrayValue
        
        //第一列tableView父视图
        headerView = UIView(frame: CGRect(x: 30, y: 50, width: 200, height: 300))
        headerView?.backgroundColor = UIColor.gray
        headerView?.isUserInteractionEnabled = true
        //设置边框，形成表格
        headerView?.layer.borderWidth = 0.5
        headerView?.layer.borderColor = UIColor.black.cgColor
        self.view.addSubview(headerView!)
        
        //可左右滑动tableView父视图
        scrollView = UIScrollView(frame: CGRect(x: 230, y: 50, width: 500, height: 300))
        scrollView?.contentSize = CGSize(width: (CGFloat)(arrX.count-1)*130, height: CGFloat(arrY.count*50))
        scrollView?.showsHorizontalScrollIndicator = true
        scrollView?.showsVerticalScrollIndicator = true
        scrollView?.bounces = false
        scrollView?.delegate = self
        scrollView?.backgroundColor = UIColor.lightGray
        //设置边框，形成表格
        scrollView?.layer.borderColor = UIColor.black.cgColor
        scrollView?.layer.borderWidth = 0.5
        self.view.addSubview(scrollView!)
        
        //第一列表
        let tableView:DataTableView = DataTableView(frame: headerView!.bounds, style: .plain)
        
        var titleArr1: [String] = []
        var levels: [String] = []
        arrY.forEach { (dic: JSON) in
            let titleStr = dic["date1"].stringValue
            titleArr1.append(titleStr)
            
            let level = dic["table_level"].stringValue
            levels.append(level)
        }
        
        tableView.titleArr = titleArr1
        tableView.levels = levels
        
        tableView.headerStr = arrX[0]["name1"].stringValue
        tableView.dele = self
        headerView?.addSubview(tableView)
        
        for i in 0 ..< arrX.count-1 {
            let tableV:DataTableView = DataTableView(frame: CGRect(x: CGFloat(130*i), y: 0, width: 130, height: 300), style: .plain)
            
            //x方向 取出key对应的字符串名字
            let xkey:String = "name\(i+2)"
            let xname:String = arrX[i+1][xkey].stringValue
            tableV.headerStr = xname
            
            //y方向
            var titleArr2: [String] = []
            for j in 0 ..< arrY.count {
                let ykey = "date\(i+2)"
                let yname = arrY[j][ykey].stringValue
                titleArr2.append(yname)
            }
            
            tableV.titleArr = titleArr2
            tableV.reloadData()
            tableV.dele = self
            scrollView?.addSubview(tableV)
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

extension ViewController: DataTableViewDelegate {
    func dataTableViewContentOffSet(contentOffSet: CGPoint) {
        for subView in scrollView?.subviews ?? [] {
            if subView is DataTableView {
                (subView as! DataTableView).setTableViewContentOffSet(contentOffset: contentOffSet)
            }
        }
        
        for subView in headerView?.subviews ?? [] {
            (subView as! DataTableView).setTableViewContentOffSet(contentOffset: contentOffSet)
        }
    }
}

extension ViewController: UIScrollViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let p:CGPoint = scrollView.contentOffset
        print(NSCoder.string(for: p))
    }
}



